File: backend/app/api/stats.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import func
from datetime import datetime
from app.db.session import get_db
from app.models.models import Transaction, Expense, Repayment, User, UserRole, Carpet
from app.services.utils import get_current_week_id
from app.api.deps import check_role, get_current_user
from app.models.models import User as UserModel

router = APIRouter()

def _weekday(day: str) -> str:
    try:
        return datetime.strptime(day, "%Y-%m-%d").strftime("%A")
    except Exception:
        return ""

def _collect_employee_stats(emp, db: Session, week_id: str = None, day: str = None):
    """Shared per-employee stats (identical for admin and employee self-service views)."""
    query = db.query(Transaction).filter(Transaction.washer_id == emp.id)
    if day:
        query = query.filter(func.date(Transaction.timestamp) == day)
    else:
        if not week_id:
            week_id = get_current_week_id()
        query = query.filter(Transaction.week_id == week_id)

    emp_revenue = query.with_entities(func.sum(Transaction.net_business_remittance)).scalar() or 0.0
    emp_wages = query.filter(Transaction.final_payout > 0).with_entities(func.sum(Transaction.final_payout)).scalar() or 0.0
    emp_shortfall = query.with_entities(func.sum(Transaction.shortfall)).scalar() or 0.0
    jobs_count = query.with_entities(func.count(Transaction.id)).scalar() or 0

    categories = query.with_entities(Transaction.category, func.count(Transaction.id)).group_by(Transaction.category).all()
    emp_counts = {cat.value: count for cat, count in categories}

    daily = db.query(
        func.date(Transaction.timestamp).label("day"),
        func.sum(Transaction.final_payout).label("wages")
    ).filter(
        Transaction.washer_id == emp.id,
        Transaction.final_payout > 0
    )
    if day:
        daily = daily.filter(func.date(Transaction.timestamp) == day)
    else:
        if not week_id:
            week_id = get_current_week_id()
        daily = daily.filter(Transaction.week_id == week_id)
    daily = daily.group_by(func.date(Transaction.timestamp)).all()
    wage_breakdown = [{"day": d.day, "day_of_week": _weekday(d.day), "wages": d.wages} for d in daily]

    debt_txns = db.query(Transaction).filter(
        Transaction.washer_id == emp.id,
        Transaction.shortfall > 0
    ).all()
    debt_sources = [{
        "id": t.id,
        "timestamp": t.timestamp.isoformat(),
        "category": t.category.value,
        "custom_category": t.custom_category,
        "expected_price": t.expected_price,
        "shortfall": t.shortfall,
        "plate_number": t.plate_number
    } for t in debt_txns]

    misc = query.with_entities(func.sum(Transaction.misc_amount)).scalar() or 0.0
    misc_txns = query.filter(Transaction.misc_amount > 0).all()
    misc_items = [{
        "id": t.id,
        "timestamp": t.timestamp.isoformat(),
        "amount": t.misc_amount,
        "description": t.misc_description,
        "category": t.category.value,
        "custom_category": t.custom_category,
        "plate_number": t.plate_number
    } for t in misc_txns]

    carpets_received = db.query(func.count(Carpet.id)).filter(
        Carpet.receiver_id == emp.id, Carpet.status == "received"
    ).scalar() or 0
    carpets_released = db.query(func.count(Carpet.id)).filter(
        Carpet.receiver_id == emp.id, Carpet.status == "released"
    ).scalar() or 0

    return {
        "id": emp.id,
        "name": emp.full_name,
        "abbreviation": emp.abbreviation,
        "jobs_count": jobs_count,
        "revenue_generated": emp_revenue,
        "wages_earned": emp_wages,
        "shortfall_total": emp_shortfall,
        "wage_breakdown": wage_breakdown,
        "current_debt": emp.debt_balance,
        "payable_balance": emp.payable_balance,
        "service_counts": emp_counts,
        "debt_sources": debt_sources,
        "misc_earned": misc,
        "misc_items": misc_items,
        "carpets_received": carpets_received,
        "carpets_released": carpets_released
    }

@router.get("/summary", dependencies=[Depends(check_role([UserRole.ADMIN, UserRole.MANAGER]))])
def get_summary(
    week_id: str = None,
    day: str = None, # YYYY-MM-DD
    db: Session = Depends(get_db)
):
    query = db.query(Transaction)
    exp_query = db.query(Expense)

    if day:
        # Filter by specific date
        query = query.filter(func.date(Transaction.timestamp) == day)
        exp_query = exp_query.filter(func.date(Expense.timestamp) == day)
    else:
        if not week_id:
            week_id = get_current_week_id()
        query = query.filter(Transaction.week_id == week_id)
        exp_query = exp_query.filter(Expense.week_id == week_id)

    total_cash = query.with_entities(func.sum(Transaction.total_paid)).scalar() or 0.0
    expenses = exp_query.with_entities(func.sum(Expense.amount)).scalar() or 0.0
    labor = query.filter(Transaction.final_payout > 0).with_entities(func.sum(Transaction.final_payout)).scalar() or 0.0
    revenue = query.with_entities(func.sum(Transaction.net_business_remittance)).scalar() or 0.0
    misc_total = query.with_entities(func.sum(Transaction.misc_amount)).scalar() or 0.0

    misc_txns = query.filter(Transaction.misc_amount > 0).order_by(Transaction.timestamp.desc()).all()
    misc_items = [{
        "id": t.id,
        "timestamp": t.timestamp.isoformat(),
        "amount": t.misc_amount,
        "description": t.misc_description,
        "category": t.category.value,
        "custom_category": t.custom_category,
        "plate_number": t.plate_number,
        "washer": f"{t.washer.full_name} ({t.washer.abbreviation})" if t.washer else None
    } for t in misc_txns]

    # Per-employee labor breakdown
    emp_labor = db.query(
        User.full_name, User.abbreviation,
        func.sum(Transaction.final_payout).label("wages")
    ).join(Transaction, User.id == Transaction.washer_id).filter(
        Transaction.final_payout > 0
    ).group_by(User.id)
    if day:
        emp_labor = emp_labor.filter(func.date(Transaction.timestamp) == day)
    else:
        emp_labor = emp_labor.filter(Transaction.week_id == week_id)
    emp_labor = emp_labor.all()
    labor_breakdown = [{"name": e.full_name, "abbreviation": e.abbreviation, "wages": e.wages} for e in emp_labor]

    # Counts by category
    categories = query.with_entities(Transaction.category, func.count(Transaction.id)).group_by(Transaction.category).all()
    category_counts = {cat.value: count for cat, count in categories}

    return {
        "week_id": week_id,
        "day": day,
        "day_of_week": _weekday(day) if day else None,
        "total_cash_received": total_cash,
        "total_expenses": expenses,
        "balance": total_cash - expenses,
        "total_labor_expense": labor,
        "total_revenue": revenue,
        "misc_total": misc_total,
        "misc_items": misc_items,
        "labor_breakdown": labor_breakdown,
        "category_counts": category_counts,
        "total_debts": db.query(func.sum(User.debt_balance)).scalar() or 0.0
    }

@router.get("/employees", dependencies=[Depends(check_role([UserRole.ADMIN, UserRole.MANAGER]))])
def get_employee_performance(
    week_id: str = None,
    day: str = None,
    db: Session = Depends(get_db)
):
    employees = db.query(User).filter(User.role == UserRole.EMPLOYEE).all()
    return [_collect_employee_stats(emp, db, week_id, day) for emp in employees]

@router.get("/employees/{employee_id}", dependencies=[Depends(check_role([UserRole.ADMIN, UserRole.MANAGER]))])
def get_employee_stats(
    employee_id: int,
    week_id: str = None,
    day: str = None,
    db: Session = Depends(get_db)
):
    emp = db.query(User).filter(User.id == employee_id).first()
    if not emp:
        raise HTTPException(status_code=404, detail="Employee not found")
    return _collect_employee_stats(emp, db, week_id, day)

@router.get("/me")
def get_my_stats(
    week_id: str = None,
    day: str = None,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_user)
):
    return _collect_employee_stats(current_user, db, week_id, day)

File: server/src/routes/stats.ts
import { Router } from 'express';
import { pool } from '../db';
import { HttpError, getCurrentWeekId, validateDay, wrap, weekdayName, weekdayNameFromDate, toUtcTimestamp } from '../utils';
import { requireRole, getCurrentUser } from '../auth';
import { ROLES } from '../types';
import type { AuthRequest } from '../auth';
import type { UserRow } from '../types';
import { iso } from '../serialize';

const router = Router();
const M = ROLES.slice(0, 2); // ADMIN, MANAGER

interface StatsScope {
  day?: string;
  week_id?: string;
}

function empScopeParams(empId: number, scope: StatsScope): { text: string; params: any[] } {
  if (scope.day) return { text: 'WHERE washer_id = $1 AND date(timestamp) = $2::date', params: [empId, scope.day] };
  const week_id = scope.week_id || getCurrentWeekId();
  return { text: 'WHERE washer_id = $1 AND week_id = $2', params: [empId, week_id] };
}

async function collectEmployeeStats(emp: UserRow, scope: StatsScope): Promise<Record<string, unknown>> {
  const q = empScopeParams(emp.id, scope);

  const [rev, wages, shortfall, jobs, cats, daily, debtSrc, misc, miscItems, received, released] = await Promise.all([
    pool.query(`SELECT COALESCE(SUM(net_business_remittance),0) AS v FROM transactions ${q.text}`, [...q.params]),
    pool.query(`SELECT COALESCE(SUM(final_payout),0) AS v FROM transactions ${q.text} AND final_payout > 0`, [...q.params]),
    pool.query(`SELECT COALESCE(SUM(shortfall),0) AS v FROM transactions ${q.text}`, [...q.params]),
    pool.query(`SELECT COUNT(id) AS v FROM transactions ${q.text}`, [...q.params]),
    pool.query(`SELECT category, COUNT(id) AS c FROM transactions ${q.text} GROUP BY category`, [...q.params]),
    pool.query(`SELECT date(timestamp) AS day, SUM(final_payout) AS wages FROM transactions ${q.text} AND final_payout > 0 GROUP BY date(timestamp)`, [...q.params]),
    pool.query(`SELECT id, timestamp, category, custom_category, expected_price, shortfall, plate_number FROM transactions ${q.text} AND shortfall > 0`, [...q.params]),
    pool.query(`SELECT COALESCE(SUM(misc_amount),0) AS v FROM transactions ${q.text}`, [...q.params]),
    pool.query(`SELECT id, timestamp, misc_amount AS amount, misc_description AS description, category, custom_category, plate_number FROM transactions ${q.text} AND misc_amount > 0`, [...q.params]),
    pool.query("SELECT COUNT(id) AS c FROM carpets WHERE receiver_id = $1 AND status = 'received'", [emp.id]),
    pool.query("SELECT COUNT(id) AS c FROM carpets WHERE receiver_id = $1 AND status = 'released'", [emp.id]),
  ]);

  const empCounts: Record<string, number> = {};
  for (const r of cats.rows) empCounts[String(r.category).toLowerCase()] = Number(r.c);

  const wageBreakdown = daily.rows.map((d) => ({
    day: toUtcTimestamp(d.day).slice(0, 10),
    day_of_week: weekdayNameFromDate(d.day),
    wages: Number(d.wages),
  }));

  const debtSources = debtSrc.rows.map((t) => ({
    id: t.id,
    timestamp: iso(new Date(t.timestamp)),
    category: String(t.category).toLowerCase(),
    custom_category: t.custom_category,
    expected_price: t.expected_price,
    shortfall: t.shortfall,
    plate_number: t.plate_number,
  }));

  const miscItemsOut = miscItems.rows.map((t) => ({
    id: t.id,
    timestamp: iso(new Date(t.timestamp)),
    amount: t.amount,
    description: t.description,
    category: String(t.category).toLowerCase(),
    custom_category: t.custom_category,
    plate_number: t.plate_number,
  }));

  return {
    id: emp.id,
    name: emp.full_name,
    abbreviation: emp.abbreviation,
    jobs_count: Number(jobs.rows[0].v),
    revenue_generated: Number(rev.rows[0].v),
    wages_earned: Number(wages.rows[0].v),
    shortfall_total: Number(shortfall.rows[0].v),
    wage_breakdown: wageBreakdown,
    current_debt: emp.debt_balance,
    payable_balance: emp.payable_balance,
    service_counts: empCounts,
    debt_sources: debtSources,
    misc_earned: Number(misc.rows[0].v),
    misc_items: miscItemsOut,
    carpets_received: Number(received.rows[0].c),
    carpets_released: Number(released.rows[0].c),
  };
}

// GET /stats/summary
router.get(
  '/summary',
  wrap(async (req: AuthRequest, res) => {
    await requireRole(req, M);
    validateDay((req.query.day as string) ?? null);
    const day = req.query.day as string | undefined;
    let week_id = req.query.week_id as string | undefined;

    const txParams: any[] = [];
    const expParams: any[] = [];
    let txConds = '';
    let expConds = '';
    if (day) {
      txParams.push(day);
      txConds = `WHERE date(timestamp) = $${txParams.length}::date`;
      expParams.push(day);
      expConds = `WHERE date(timestamp) = $${expParams.length}::date`;
    } else {
      if (!week_id) week_id = getCurrentWeekId();
      txParams.push(week_id);
      txConds = `WHERE week_id = $${txParams.length}`;
      expParams.push(week_id);
      expConds = `WHERE week_id = $${expParams.length}`;
    }

    const [totalCash, expenses, labor, revenue, miscTotal, miscRows, catRows, totalDebt, empLabor] = await Promise.all([
      pool.query(`SELECT COALESCE(SUM(total_paid),0) AS v FROM transactions ${txConds}`, txParams),
      pool.query(`SELECT COALESCE(SUM(amount),0) AS v FROM expenses ${expConds}`, expParams),
      pool.query(`SELECT COALESCE(SUM(final_payout),0) AS v FROM transactions ${txConds} AND final_payout > 0`, txParams),
      pool.query(`SELECT COALESCE(SUM(net_business_remittance),0) AS v FROM transactions ${txConds}`, txParams),
      pool.query(`SELECT COALESCE(SUM(misc_amount),0) AS v FROM transactions ${txConds}`, txParams),
      pool.query(
        `SELECT id, timestamp, misc_amount AS amount, misc_description AS description, category, custom_category, plate_number, washer_id
         FROM transactions ${txConds} AND misc_amount > 0 ORDER BY timestamp DESC`,
        txParams,
      ),
      pool.query(`SELECT category, COUNT(id) AS c FROM transactions ${txConds} GROUP BY category`, txParams),
      pool.query(`SELECT COALESCE(SUM(debt_balance),0) AS v FROM users`),
      pool.query(
        `SELECT users.full_name, users.abbreviation, SUM(transactions.final_payout) AS wages
         FROM users JOIN transactions ON users.id = transactions.washer_id
         WHERE transactions.final_payout > 0 AND ${day ? 'date(transactions.timestamp) = $1::date' : 'transactions.week_id = $1'}
         GROUP BY users.id`,
        day ? [day] : [week_id],
      ),
    ]);

    // Resolve washer names for misc_items
    const washerIds = [...new Set(miscRows.rows.map((r: any) => r.washer_id).filter(Boolean))];
    const nameOf: Record<number, string> = {};
    if (washerIds.length) {
      const users = (
        await pool.query(`SELECT id, full_name, abbreviation FROM users WHERE id = ANY($1::int[])`, [washerIds])
      ).rows;
      for (const u of users) nameOf[u.id] = `${u.full_name} (${u.abbreviation})`;
    }

    const categoryCounts: Record<string, number> = {};
    for (const r of catRows.rows) categoryCounts[String(r.category).toLowerCase()] = Number(r.c);

    const miscItems = miscRows.rows.map((t: any) => ({
      id: t.id,
      timestamp: iso(new Date(t.timestamp)),
      amount: t.amount,
      description: t.description,
      category: String(t.category).toLowerCase(),
      custom_category: t.custom_category,
      plate_number: t.plate_number,
      washer: t.washer_id ? nameOf[t.washer_id] ?? null : null,
    }));

    const totalCashV = Number(totalCash.rows[0].v);
    const expensesV = Number(expenses.rows[0].v);
    const laborV = Number(labor.rows[0].v);
    res.json({
      week_id,
      day: day ?? null,
      day_of_week: day ? weekdayName(day) : null,
      total_cash_received: totalCashV,
      total_expenses: expensesV,
      balance: totalCashV - expensesV,
      total_labor_expense: laborV,
      total_revenue: Number(revenue.rows[0].v),
      misc_total: Number(miscTotal.rows[0].v),
      misc_items: miscItems,
      labor_breakdown: empLabor.rows.map((e: any) => ({
        name: e.full_name,
        abbreviation: e.abbreviation,
        wages: Number(e.wages),
      })),
      category_counts: categoryCounts,
      total_debts: Number(totalDebt.rows[0].v),
    });
  }),
);

// GET /stats/employees
router.get(
  '/employees',
  wrap(async (req: AuthRequest, res) => {
    await requireRole(req, M);
    validateDay((req.query.day as string) ?? null);
    const scope: StatsScope = {
      day: req.query.day as string | undefined,
      week_id: req.query.week_id as string | undefined,
    };
    const employees = (
      await pool.query("SELECT * FROM users WHERE role = 'EMPLOYEE' ORDER BY id")
    ).rows as UserRow[];
    const out = [];
    for (const emp of employees) out.push(await collectEmployeeStats(emp, scope));
    res.json(out);
  }),
);

// GET /stats/employees/:id
router.get(
  '/employees/:id',
  wrap(async (req: AuthRequest, res) => {
    await requireRole(req, M);
    validateDay((req.query.day as string) ?? null);
    const id = Number(req.params.id);
    if (!Number.isInteger(id)) throw new HttpError(404, 'Employee not found');
    const emp = (
      await pool.query('SELECT * FROM users WHERE id = $1', [id])
    ).rows[0] as UserRow | undefined;
    if (!emp) throw new HttpError(404, 'Employee not found');
    const scope: StatsScope = {
      day: req.query.day as string | undefined,
      week_id: req.query.week_id as string | undefined,
    };
    res.json(await collectEmployeeStats(emp, scope));
  }),
);

// GET /stats/me
router.get(
  '/me',
  wrap(async (req: AuthRequest, res) => {
    const user = await getCurrentUser(req);
    validateDay((req.query.day as string) ?? null);
    const scope: StatsScope = {
      day: req.query.day as string | undefined,
      week_id: req.query.week_id as string | undefined,
    };
    res.json(await collectEmployeeStats(user, scope));
  }),
);

export default router;